system-tests/lib/cre/topology.go

```diff
@@ -0,0 +1,102 @@
+package cre
+
+import (
+	"fmt"
+
+	libc "github.com/smartcontractkit/chainlink/system-tests/lib/conversions"
+
+	"github.com/smartcontractkit/chainlink/system-tests/lib/infra"
+)
+
+const (
+	OCRPeeringPort          = 5001
+	CapabilitiesPeeringPort = 6690
+)
+
+var (
+	NodeTypeKey            = "type"
+	HostLabelKey           = "host"
+	IndexKey               = "node_index"
+	ExtraRolesKey          = "extra_roles"
+	NodeIDKey              = "node_id"
+	NodeOCR2KeyBundleIDKey = "ocr2_key_bundle_id"
+	NodeP2PIDKey           = "p2p_id"
+	DONIDKey               = "don_id"
+	EnvironmentKey         = "environment"
+	ProductKey             = "product"
+	DONNameKey             = "don_name"
+)
+
+type Topology struct {
+	WorkflowDONID           uint64                  `toml:"workflow_don_id" json:"workflow_don_id"`
+	DonsMetadata            []*DonMetadata          `toml:"dons_metadata" json:"dons_metadata"`
+	CapabilitiesPeeringData CapabilitiesPeeringData `toml:"capabilities_peering_data" json:"capabilities_peering_data"`
+	OCRPeeringData          OCRPeeringData          `toml:"ocr_peering_data" json:"ocr_peering_data"`
+	GatewayConnectorOutput  *GatewayConnectorOutput `toml:"gateway_connector_output" json:"gateway_connector_output"`
+}
+
+func NewTopology(nodeSetInput []*CapabilitiesAwareNodeSet, infraInput infra.Provider) (*Topology, error) {
+	// TODO this setup is awkward, consider an withInfra opt to constructor
```

Comment:
but infra isn't optional, so why would we pass it as an option? we need it to generate certain URLs later used during configuration generation.